### ISSUE TITLE
tridactyl sometimes fails to source RC file

### ISSUE THREAD

**Issue by buckley-w-david** [tridactyl/tridactyl#2001]
Tridactyl version: 1.17.1pre3326-e5c676a
Firefox version: Mozilla Firefox 70.0.1
Operating system: linux

*   Brief description of the problem: tridactyl sometimes fails to source RC file

*   Config (in a new tab, run `:viewconfig`, copy the url and paste it somewhere like gist.github.com): https://gist.github.com/buckley-w-david/0fd6227d1e3f213a83b3df4b1c45335c#file-config-json

*   Contents of ~/.tridactylrc or ~/.config/tridactyl/tridactylrc (if they exist): https://gist.github.com/buckley-w-david/0fd6227d1e3f213a83b3df4b1c45335c#file-tridactylrc

For the last little while (not sure how long, but it's probably been more than a week), I've been having a strange issue where I'm noticing that my RC file isn't being sourced. My bindings aren't present, my theme is wrong, etc. When this happens I am just manually sourcing it which works fine.

Not too big a deal since I can manually source, but figured since it persisted for a little while I'd report it.

**Comment by ExpandingMan** [comment]
Tridactyl version: 1.17.1pre3304-7dcf183
Firefox version: Mozilla Firefox 71.0
Operating system: Manjaro Linux

I'm completely unable to get tridactyl to source the `tridactylrc`.  It doesn't do it automatically, it doesn't do it with `source`, and it doesn't do it with `source path/to/tridactylrc`.

For the time being I'm just using a very simple config from one of the examples
```
" vim: filetype=vim

" This wipes all existing settings. This means that if a setting in this file
" is removed, then it will return to default. In other words, this file serves
" as an enforced single point of truth for Tridactyl's configuration.
" sanitize tridactyllocal tridactylsync

" Just use a blank page for new tab. It would be nicer to use the standard
" Firefox homepage, but Tridactyl doesn't support this yet.
set newtab about:blank

" Ctrl-F should use the browser's native 'find' functionality.
unbind <C-f>

" But also support Tridactyl search too.
bind / fillcmdline find
bind ? fillcmdline find -?
bind n findnext 1
bind N findnext -1
" Remove search highlighting.
" bind ,<Space> nohlsearch
" Use sensitive case. Smart case would be nice here, but it doesn't work.
" set findcase sensitive

" Smooth scrolling, yes please. This is still a bit janky in Tridactyl.
set smoothscroll true

" The default jump of 10 is a bit much.
bind j scrollline 5
bind k scrollline -5

" K and J should move between tabs. x should close them.
bind K tabprev
bind J tabnext

bind O tabopen

" Don't run Tridactyl on some web sites because it doesn't work well, or
" because the web site has its own keybindings.
" autocmd DocStart mail.google.com mode ignore

" Sometimes the status bar in the bottom left corner overlaps the Tridactyl
" command line, so set an option to move the status bar to the right.
guiset_quiet hoverlink right
```

**Comment by alerque** [comment]
I can't solve your sourcing issue, but as long as you are editing your config you might get more miliage out of a a specific file type. Change in the hack:

```vim
" vim: filetype=vim
```

…for [vim-tridactyl](https://github.com/tridactyl/vim-tridactyl) and:

```vim
" vim: filetype=tridactyl
```

**Comment by ExpandingMan** [comment]
Thanks, I wish I could get it to source so that would actually work :laughing:

**Comment by bovine3dom** [comment]
@ExpandingMan FWIW, we offer a version of Tridactyl with no new tab page here: https://tridactyl.cmcaine.co.uk/betas/nonewtab/tridactyl_no_new_tab_beta-latest.xpi

To debug your issue, could you try making a really simple RC file (e.g. just `tabclose`), sourcing it and seeing if that works?

**Comment by alerque** [comment]
@ExpandingMan As a little more verbose example, I saved the following to `~/scratch/tritest`:
```
tabnew example.com
```

Then sourced in Tridactyl:

```
source ~/scratch/tritest
```

...which works for me right now and instantly pops open a new tab. I have had issues with `source` in the past though, so some bug is there, but a bit elusive.

**Comment by ExpandingMan** [comment]
Sorry guys, the `nativeinstall` script had failed to run properly the first time, and that's why `source` wasn't working for me.

By the way (sorry if this isn't the correct place to ask this) but I can't for the life of me figure out how to do a bind command that then takes an argument (like `bind O tabopen $type_command_here`), any help on that?

**Comment by glacambre** [comment]
@ExpandingMan Do you mean something like `bind O fillcmdline tabopen`?

**Comment by ExpandingMan** [comment]
That's it, thanks!  I had the arguments in the wrong order.

**Comment by Mange** [comment]
I too have issues with my RC file not being loaded automatically on Firefox getting restarted, but a manual `:source` always works.

Tridactyl 1.17.1pre3501-b6e58c4
Firefox Developer Edition 73.0b9

```
› pacman -Qi firefox-tridactyl-native | grep -E 'Install Date|Version'
Version         : 1.17.1-1
Install Date    : Mon 03 Feb 2020 11:48:03 AM CET
```

Can I do something else to help out debugging this?

**Comment by bovine3dom** [comment]
Could you check that `viewconfig autocmds` has a TriStart key with `*.*: source_quiet`? Another thing to do is to make your RC file very simple (e.g. just a single `tabopen news.bbc.co.uk`) so we can see if it is being called at all.

**Comment by Mange** [comment]
```json
{
  "DocStart": {
    "smile.reddit.com": "js tri.excmds.urlmodify(\"-t\", \"smile\", \"old\")",
    "smile.amazon.co.uk": "js tri.excmds.urlmodify(\"-t\", \"www\", \"smile\")",
    "hemnet.se": "js tri.excmds.urlmodify(\"-t\", \"hement\", \"hemnet\")"
  },
  "DocLoad": {
    "^https://github.com/tridactyl/tridactyl/issues/new$": "issue"
  },
  "DocEnd": {},
  "TriStart": {
    ".*": "source_quiet"
  },
  "TabEnter": {},
  "TabLeft": {},
  "FullscreenChange": {},
  "FullscreenEnter": {},
  "FullscreenLeft": {}
}
```

It's only named `.*`, not `*.*`. Could that be a hint?

I tried running `:source_quiet` and it loaded the config successfully.

> Another thing to do is to make your RC file very simple (e.g. just a single `tabopen news.bbc.co.uk`) so we can see if it is being called at all.

Do we think it might be the contents of it that causes problems? Because I can clearly tell when it's loaded or not because I don't use the light color theme, so I only have to look at the modeline to tell. If we only want a smaller file to tell if it's loaded or not then it should not be needed I think.

**Comment by bovine3dom** [comment]
The .* is weird. Try `autocmd TriStart *.* source_quiet` and then restart :)

**Comment by glacambre** [comment]
`.*` is normal (see https://github.com/tridactyl/tridactyl/blob/266988344ec8fcd524311e61914d82ff44fcc703/src/background.ts#L85 and https://github.com/tridactyl/tridactyl/blob/f179f95880a708d9c09e7b46b4484e4fc5a1edc2/src/lib/config.ts#L395). `*.*` would be weird.

**Comment by bovine3dom** [comment]
Oh, you're right. Thanks glacambre. " and * are hard to distinguish on the terminal on my phone apparently. I had looked it up to check about 3 times and read it wrong every time. :(

In which case your issue sounds like the original one for which I don't really know the answer.

You might try shortening your RC file to test to see if it loads then but it seems unlikely if none of it is being run at all.

**Comment by Mange** [comment]
I tried to restart Firefox a few times. It's definitely not consistent.


1. OK
2. OK
3. OK, but took a few seconds before it loaded
4. Did not load
5. Did not load
6. Did not load
7. OK
8. OK
9. OK
10. OK

I stopped here and opened this issue to type out the comment. Very strange.

**Comment by bovine3dom** [comment]
Ok, that's great - it is the same issue as the OP. I'll have a think about this and ping you if a need s guinea pig. Thanks for investigating it further :)

**Comment by rrajath** [comment]
Hi,
I'm facing a similar issue. I have a set of overrides like using a different key to close a tab, and simple keybinding changes like that. I've stored this config at `~/.tridactylrc`. Every time I open a new tab, Tridactyl resets the config - meaning, my overrides no longer work. If I run `:source` in tridactyl console, it'll apply my overrides to that page. But as soon as I open a new page, it's all reset again.

I'm using the latest version of Firefox on OSX and the latest version of Tridactyl.

Can someone please help me out?

**Comment by Federico-Ciuffardi** [comment]
`set storageloc local` worked for me.

**Comment by bovine3dom** [comment]
@Federico-Ciuffardi what version of Tridactyl are you using? If it's a recent version, I'm afraid that that setting is a placebo : )

**Comment by Federico-Ciuffardi** [comment]
1.20.2
that's weird, like 100% placebo? because it actually fixed my problem.

**Comment by bovine3dom** [comment]
Yup, the only references to `storageloc` are places where we are updating old configs: https://github.com/tridactyl/tridactyl/blob/eca1e1614b8d0e51dcb7369a3b4026fed72ad209/src/lib/config.ts#L1586

I'm glad your problem has gone away : )
